repurpose

danielnorberg opened the pull request:
- Monitor manifest file(s)
- Recreate containers in response to manifest changes
- Kill containers in namespace that are not listed in manifest file(s)
- Internalize container keep-alive loop so that killing the container-agent lets you kill the containers.
- Use standard logging lib.
- Use docker client from docker-stress
- Remove infrastructure containers and container network stack sharing
- Remove manifest version check given substantial capability & behavior divergence
- Log errors instead of exiting on configuration validation failures.

Commit:
fix setup.py

Commit:
repurpose all the things

* Regularly reload config file
* Recreate containers in response to config changes
* Kill containers in namespace that are not listed in config file
* Use standard logging lib

philipcristiano commented on container_agent/docker_client.py:
```diff
@@ -0,0 +1,132 @@
+DEFAULT_DOCKER_ENDPOINT = environ.get('DOCKER_HOST', 'unix:///var/run/docker.sock')
+
+try:
+    p = Popen('which docker', stdout=PIPE, stderr=PIPE, shell=True)
```
Trying to find the location of docker should be moved to the **init** for the docker client and be called in the event the client isn't specified. Making syscalls in the root of the module isn't good, any module that imports this will trigger a syscall. Slowing down the init isn't great but it's better than calling it because of an import

philipcristiano commented on container_agent/docker_client.py:
```diff
@@ -0,0 +1,132 @@
+    if p.returncode:
+        raise Exception()
+    DEFAULT_DOCKER_CLI = out.strip()
+except:
```
You should be specific about what exceptions should be caught, especially since you raise a general exception above, you might just want to set DEFAULT_DOCKER_CLI above and use that as the else of the conditional as well as expected errors

danielnorberg replied:
not sure how you want this to look

philipcristiano replied:
Probably a method on the client called the first time a client is created without an argument, something like

```
def _detect_system_cli(klass):
     location = '/usr/bin/docker'
     try:
          p = Popen('which docker', stdout=PIPE, stderr=PIPE, shell=True)
     except (whatever popen can throw):
          logging.info('error detecting docker cli location, using default of {0}'.format(location))
     else:
         out, err = p.communicate()
         if p.returncode == 0:
           location = out.strip()
     klass._system_cli = location
```

philipcristiano commented on container_agent/docker_client.py:
```diff
@@ -0,0 +1,132 @@
+class CliDockerClient(object):
+    def __init__(self, docker=None, endpoint=None):
+        super(CliDockerClient, self).__init__()
+        self.docker = docker or DEFAULT_DOCKER_CLI
```
This will trigger on falsey values passed through on docker, while not valid you should at least honor the arguments passed through and check if `docker is None` instead of checking the truthyness

danielnorberg replied:
done

Commit:
fix tests and pep8 errors

philipcristiano commented on container_agent/docker_client.py:
```diff
@@ -0,0 +1,143 @@
+        self.cli_check('rm', container_id)
+
+    def list_containers(self, needle=''):
+        if not needle:
```
This could be written without the `not` and use the needle if one existed instead of `not` first.

Commit:
client: stricter default arg handling

philipcristiano commented on container_agent/run_containers.py:
```diff
@@ -26,28 +26,28 @@
-  - Docker 0.11 or higher (for the --net flag)
-  - Docker daemon runs with =r=false (for safer restart behavior)
+  - Docker 0.11 or higher
+  - Docker daemon runs with -r=false (for safer restart behavior)
```
Whitespace changes should be in a separate commit

danielnorberg replied:
it's not whitespace changes

philipcristiano replied:
apparently I just don't understand when GH does highlighting for character changes vs line changes

Commit:
fix except syntax

Commit:
take a glob of config files

Merge multiple config files and run all containers. This allows us to
e.g. use puppet to drop manifests into place as files and have
container-agent pick up and run them. To stop running these containers
it is then sufficient to remove the manifest file.

Also remove the version check as we're diverging significantly from
upstream anyway

Commit:
more robust docker cli detection

Commit:
docstrings

philipcristiano commented:
+1
